File: files/web/serial.js
/*****************************************************************************
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
*****************************************************************************/


"use strict";


import {tools, $} from "../tools.js";
import {wm} from "../wm.js";


export function Serial() {
	var self = this;

	/************************************************************************/

	var __ws = null;
	var __state = null;
	var __term = null;
	var __fit_addon = null;
	var __term_container = null;
	var __initialized = false;
	var __pending_data = "";
	var __resize_observer = null;

	var __init__ = function() {
		__term_container = $("serial-term-container");

		tools.el.setOnClick($("serial-clear-button"), function() {
			if (__term) {
				__term.clear();
			}
		});

		let autoscroll_switch = $("serial-autoscroll-switch");
		if (autoscroll_switch) {
			tools.storage.bindSimpleSwitch(autoscroll_switch, "serial.autoscroll", true);
		}

		let speed_select = $("serial-speed-select");
		if (speed_select) {
			speed_select.addEventListener("change", function() {
				tools.httpPost("api/serial/set_speed", {"speed": speed_select.value}, function(http) {
					if (http.status !== 200) {
						wm.error("Failed to set baud rate", http.responseText);
					}
				});
			});
		}

		// Initialize xterm when the serial window becomes visible
		let serial_window = $("serial-window");
		if (serial_window) {
			let observer = new MutationObserver(function() {
				if (!serial_window.classList.contains("hidden") && !__initialized) {
					__initTerm();
				}
				if (!serial_window.classList.contains("hidden") && __term && __fit_addon) {
					setTimeout(function() { __fit_addon.fit(); }, 50);
				}
			});
			observer.observe(serial_window, {attributes: true, attributeFilter: ["class"]});
		}
	};

	var __initTerm = function() {
		if (__initialized || !__term_container) {
			return;
		}
		if (typeof Terminal === "undefined") {
			return;
		}

		__term = new Terminal({
			cursorBlink: true,
			cursorStyle: "block",
			fontSize: 14,
			fontFamily: "'Courier New', monospace",
			theme: {
				background: "#1e1e1e",
				foreground: "#d4d4d4",
				cursor: "#d4d4d4",
				selectionBackground: "#444444",
			},
			scrollback: 5000,
			convertEol: false,
			allowProposedApi: true,
		});

		if (typeof FitAddon !== "undefined") {
			__fit_addon = new FitAddon.FitAddon();
			__term.loadAddon(__fit_addon);
		}

		__term.open(__term_container);

		if (__fit_addon) {
			__fit_addon.fit();
		}

		// Send keypresses to serial
		__term.onData(function(data) {
			if (__ws && __ws.readyState === WebSocket.OPEN) {
				__ws.send(JSON.stringify({
					"event_type": "serial_write",
					"event": {"data": data},
				}));
			}
		});

		// Resize terminal when container resizes
		__resize_observer = new ResizeObserver(function() {
			if (__fit_addon && __term) {
				__fit_addon.fit();
			}
		});
		__resize_observer.observe(__term_container);

		__initialized = true;

		// Flush any data that arrived before terminal was ready
		if (__pending_data) {
			__term.write(__pending_data);
			__pending_data = "";
		}
	};

	/************************************************************************/

	self.setSocket = function(ws) {
		__ws = ws;
	};

	self.setState = function(state) {
		if (state) {
			if (!__state) {
				__state = {};
			}
			if (state.enabled !== undefined) {
				__state.enabled = state.enabled;
				tools.feature.setEnabled($("serial-dropdown"), __state.enabled);
			}
			if (state.online !== undefined) {
				__state.online = state.online;
				$("serial-led").className = (state.online ? "led-green" : "led-gray");
				$("serial-led").title = (state.online ? "Serial: Connected" : "Serial: Disconnected");
			}
			if (state.device !== undefined) {
				__state.device = state.device;
				let el = $("serial-device-info");
				if (el) {
					el.innerText = state.device;
				}
			}
			if (state.speed !== undefined) {
				__state.speed = state.speed;
				let speed_select = $("serial-speed-select");
				if (speed_select && speed_select !== document.activeElement) {
					speed_select.value = String(state.speed);
				}
			}
			if (state.data !== undefined && state.data.length > 0) {
				if (__term) {
					__term.write(state.data);
				} else {
					__pending_data += state.data;
				}
			}
		} else {
			__state = null;
			$("serial-led").className = "led-gray";
			$("serial-led").title = "Serial: Disconnected";
			tools.feature.setEnabled($("serial-dropdown"), false);
		}
	};

	__init__();
}

File: files/api/serial.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


from aiohttp.web import Request
from aiohttp.web import Response

from ....htserver import exposed_http
from ....htserver import exposed_ws
from ....htserver import make_json_response
from ....htserver import WsSession

from ....plugins.serial import BaseSerial

from ....validators.basic import valid_bool
from ....validators.hw import valid_tty_speed


# =====
class SerialApi:
    def __init__(self, serial: BaseSerial) -> None:
        self.__serial = serial

    # =====

    @exposed_http("GET", "/serial")
    async def __state_handler(self, _: Request) -> Response:
        return make_json_response(await self.__serial.get_state())

    @exposed_http("POST", "/serial/write")
    async def __write_handler(self, req: Request) -> Response:
        data = await req.text()
        await self.__serial.write(data)
        return make_json_response()

    @exposed_http("POST", "/serial/set_speed")
    async def __set_speed_handler(self, req: Request) -> Response:
        speed = valid_tty_speed(req.query.get("speed"))
        await self.__serial.set_speed(speed)
        return make_json_response()

    @exposed_http("POST", "/serial/upload")
    async def __upload_handler(self, req: Request) -> Response:
        address_str = req.query.get("address")
        if not address_str:
            return make_json_response({"error": "address parameter required"}, status=400)
        address = int(address_str, 16)
        data = await req.read()
        if not data:
            return make_json_response({"error": "no data in request body"}, status=400)
        call = valid_bool(req.query.get("call", False))
        go = valid_bool(req.query.get("go", False))
        result = await self.__serial.upload(address, data, call=call, go=go)
        if "error" in result:
            return make_json_response(result, status=409)
        return make_json_response(result)

    @exposed_ws("serial_write")
    async def __ws_write_handler(self, _: WsSession, event: dict) -> None:
        data = event.get("data", "")
        if data:
            await self.__serial.write(data)
